utils/apiclient/client.go

```diff
@@ -0,0 +1,85 @@
+package apiclient
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"time"
+
+	"github.com/pkg/errors"
+)
+
+const (
+	maxRetries     = 5
+	initialBackoff = 1 * time.Second
+)
+
+func isRetryableStatusCode(statusCode int) bool {
+	return statusCode == http.StatusTooManyRequests || statusCode == http.StatusServiceUnavailable
+}
+
+func (c *APIClient) GetURL(endpoint string, qstr url.Values) string {
+	return fmt.Sprintf("%s/%s?%s", c.BaseURL, endpoint, qstr.Encode())
+}
+
+func (c *APIClient) CallAPI(reqParams RequestParams) (interface{}, error) {
+	if reqParams.QueryParams == nil {
+		reqParams.QueryParams = url.Values{}
+	}
+
+	if reqParams.Headers == nil {
+		reqParams.Headers = map[string]interface{}{}
+	}
+
+	url := c.GetURL(reqParams.Endpoint, reqParams.QueryParams)
+	reqBody, err := CreateRequestBody(reqParams.RequestType, url)
+	if err != nil {
+		return nil, errors.Wrap(err, "http request creation failed")
+	}
+
+	SetAuthHeaders(reqBody, c.AuthType, c.AuthHeaders)
+	SetHeaders(reqBody, reqParams.Headers)
+	client := c.HTTP
+	if client == nil {
+		client = &http.Client{}
+	}
+
+	var result interface{}
+	retries := 0
+
+	for retries <= maxRetries {
+		resp, err := client.Do(reqBody)
+		if err != nil {
+			return nil, errors.Wrap(err, "http request failed")
+		}
+		defer resp.Body.Close()
+
+		if resp.StatusCode == http.StatusOK {
+			body, err := ioutil.ReadAll(resp.Body)
+			if err != nil {
+				return nil, fmt.Errorf("failed to read response body: %w", err)
+			}
+
+			if err := json.Unmarshal(body, &result); err != nil {
+				return nil, fmt.Errorf("failed to unmarshal JSON: %w", err)
```

Comment:
nit: I think these two errors can use the same `errors.Wrap()` as above instead of `fmt.Errorf()` and passing it the err

Reply:
Do you mean following change?
```suggestion
				return nil, fmt.Errorf("failed to read response body: %w", err)
			}

			if err := json.Unmarshal(body, &result); err != nil {
				return nil, err
```

Reply:
Addressed in https://github.com/stellar/go/pull/5534/commits/f5792cd3a40b6d35cab7367424685caa69fa6931

Reply:
Sorry I meant that instead of doing `fmt.Errorf("failed to read response body: %w", err)` you could just do `errors.Wrap(err, "supplemental error text")`

But it looks like I got the recommendation wrong. It seems like we want to avoid using `errors.Wrap()` in favor of using `fmt.Errorf()`

Two notable things from this stack overflow question:
https://stackoverflow.com/questions/61933650/whats-the-difference-between-errors-wrapf-errors-errorf-and-fmt-errorf
"*The github.com/pkg/errors package is no longer actively maintained. The repo has been archived, and is no longer accepting updates, but remains available for use."
"One difference between Wrapf(err, "") and Errorf("%w", err) is that wrapf would return nil if err is nil, but errorf would return an error regardless."

Reply:
Hrm. This is interesting. It makes sense to use `fmt.errorf` to get the error regardless.

Reply:
Addressed in https://github.com/stellar/go/pull/5534/commits/66f236c8f1c2fb2a72c43ae20a315c44b8d15291